### PR TITLE
Add generation module from #106 (4/6)

### PR DESCRIPTION
**Pull request by irinakhismatullina**
Module for candidate generation. Small examples of usage can be found in `test_generation.py`.

In a few words generation process works like this:
- Takes dataset of tokens to check with their context. 
- For every token from the dataset generates candidates  for correction (closest on edit distance to the tokens and most frequent & closest on embeddings cosine distance to the token and context).
- For every pair `(checked_token, candidate)` a list of features is generated. Edit distance, cosine distance between and vectors of checked token, candidate, context are taken.
- The dataset of candidates for every checked token and corresponding features is returned.

Ready to review and merge. Based on #109.

### CODE REVIEW

**Comment by zurk** [comment]
image
wow :)

**Comment by irinakhismatullina** [comment]
Keep calm, it's data:)

File: lookout/style/typos/generation.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+        self.radius = radius
+
+        self.tokens = read_vocabulary(vocabulary_file)
+        self.frequencies = read_frequencies(frequencies_file)
+
+    def generate_candidates(self, data: pandas.DataFrame, threads_number: int,
+                            save_candidates_file: str = None,
+                            start_pool_size: int = 64) -> pandas.DataFrame:
+        """
+        Generates candidates for typos inside data.
```
Comment (vmarkovtsev) [inline lookout/style/typos/generation.py]: New line after the desc

File: lookout/style/typos/generation.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+        self.neighbors_number = self.DEFAULT_NEIGHBORS_NUMBER
+        self.taken_for_distance = self.DEFAULT_TAKEN_FOR_DISTANCE
+        self.max_distance = self.DEFAULT_MAX_DISTANCE
+        self.radius = self.DEFAULT_RADIUS
+        self.tokens = []
+        self.frequencies = {}
+
+    def construct(self, vocabulary_file: str, frequencies_file: str, embeddings_file: str,
+                  neighbors_number: int = DEFAULT_NEIGHBORS_NUMBER,
+                  taken_for_distance: int = DEFAULT_TAKEN_FOR_DISTANCE,
```
Comment (vmarkovtsev) [inline lookout/style/typos/generation.py]: taken -> token

File: lookout/style/typos/generation.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+        if save_candidates_file is not None:
+            candidates.to_pickle(save_candidates_file)
+
+        return candidates
+
+    def _lookup_corrections_for_token(self, typo_info: TypoInfo) -> List[List[Union[str, float]]]:
+        candidates = []
+        candidate_tokens = self._get_candidate_tokens(typo_info)
+
```
Comment (vmarkovtsev) [inline lookout/style/typos/generation.py]: Redundant empty lines

File: lookout/style/typos/generation.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+                                if candidate in self.tokens])
+        if not len(candidate_tokens):
+            candidate_tokens.add(typo_info.typo)
+        return candidate_tokens
+
+    def _generate_features(self, typo_info: TypoInfo, dist: int, typo_vec: numpy.ndarray,
+                           candidate: str, candidate_vec: numpy.ndarray
+                           ) -> List[Union[str, float]]:
+        """
+        Features for correction candidate.
```
Comment (vmarkovtsev) [inline lookout/style/typos/generation.py]: New line after the desc

Comment (vmarkovtsev) [inline lookout/style/typos/generation.py]: Compile features for a single correction candidate

File: lookout/style/typos/generation.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        checker_tree["_deletes"] = {int(h): deletes
+                                    for h, deletes in checker_tree["_deletes"].items()}
+        self.checker.__dict__.update(checker_tree)
+
+    def __str__(self):
+        return ("Vocabulary_size %d. \n"
+                "Neighbors number %d. \n"
+                "Maximum distance for search %d. \n"
+                "Maximum distance allowed %d. \n"
+                "Taken for distance %d.") % (len(self.tokens), self.neighbors_number,
```
Comment (vmarkovtsev) [inline lookout/style/typos/generation.py]: Taken -> Token

File: lookout/style/typos/generation.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+                                             self.max_distance, self.radius,
+                                             self.taken_for_distance)
+
+
+def get_candidates_features(candidates: pandas.DataFrame) -> numpy.ndarray:
+    return candidates.drop(columns=[ID_COLUMN, TYPO_COLUMN,
+                                    CANDIDATE_COLUMN]).as_matrix().astype(float)
```
Comment (vmarkovtsev) [inline lookout/style/typos/generation.py]: better: float -> numpy.float32

File: lookout/style/typos/symspell.py [hunk 7]
```diff
@@ -146,7 +155,7 @@ def create_dictionary(self, corpus):
             corpus (str): Path to corpus file.
         """
         if os.path.exists(corpus):
-            with open(corpus, 'r') as f:
+            with open(corpus, "r") as f:
```
Comment (vmarkovtsev) [inline lookout/style/typos/symspell.py]: `"r"` not needed

**Pull request merged**